sw/device/lib/base/print.c

```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/print.h"
+
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+// Standard format specifiers.
+static const char kPercent = '%';
+static const char kCharacter = 'c';
+static const char kString = 's';
+static const char kSignedDec1 = 'd';
+static const char kSignedDec2 = 'i';
+static const char kUnsignedOct = 'o';
+static const char kUnsignedHexLow = 'x';
+static const char kUnsignedHexHigh = 'X';
+static const char kUnsignedDec = 'u';
+static const char kCharsSoFar = 'n';
+static const char kPointer = 'p';
+
+// Verilog-style format specifiers.
+static const char kSvBinary = 'b';
+static const char kSvHexLow = 'h';
+static const char kSvHexHigh = 'H';
+
+static const char kSizedStr = 'z';
+
+// NOTE: all of the lengths of the strings below are given so that the NUL
+// terminator is left off; that way, |sizeof(kConst)| does not include it.
+static const char kDigitsLow[16] = "0123456789abcdef";
+static const char kDigitsHigh[16] = "0123456789ABCDEF";
+
+static const char kErrorNul[17] = "%<unexpected nul>";
+static const char kUnknownSpec[15] = "%<unknown spec>";
+
+static size_t base_dev_null(void *data, const char *buf, size_t len) {
+  return len;
+}
+static buffer_sink_t base_stdout = {
+  .data = NULL,
+  .sink = &base_dev_null,
+};
+
+void base_set_stdout(buffer_sink_t out) {
+  if (out.sink == NULL) {
+    out.sink = &base_dev_null;
+  }
+  base_stdout = out;
+}
+
+size_t base_printf(const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+  return base_vfprintf(base_stdout, format, args);
+}
+
+typedef struct snprintf_captures_t {
+  char *buf;
+  size_t bytes_left;
+} snprintf_captures_t;
+
+static size_t snprintf_sink(void *data, const char *buf, size_t len) {
+  snprintf_captures_t *captures = (snprintf_captures_t *)data;
+  if (captures->bytes_left == 0) {
+    return 0;
+  }
+
+  if (len > captures->bytes_left) {
+    len = captures->bytes_left;
+  }
+  __builtin_memcpy(captures->buf, buf, len);
+  captures->buf += len;
+  captures->bytes_left += len;
+  return len;
+}
+
+size_t base_snprintf(char *buf, size_t len, const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+
+  snprintf_captures_t data = {
+    .buf = buf,
+    .bytes_left = len,
+  };
+  buffer_sink_t sink = {
+    .data = &data,
+    .sink = &snprintf_sink,
+  };
+  return base_vfprintf(sink, format, args);
+}
+
+size_t base_fprintf(buffer_sink_t out, const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+  return base_vfprintf(out, format, args);
+}
+
+static size_t write_digits(buffer_sink_t out, uint32_t value, uint32_t width,
+                           uint32_t base, const char *glyphs) {
+  char buffer[32];
+  size_t len = 0;
+  while (value > 0) {
+    uint32_t digit = value % base;
+    value /= base;
+    buffer[31 - len] = glyphs[digit];
+    ++len;
+  }
+  width = width == 0 ? 1 : width;
```

Comment:
One of your own suggestions on a different change was to avoid "?" operator.

Reply:
The ternary is only a problem when it spans many lines. Something like a clamping operation should be a ternary.